docs/RFCS/20200810_leaseholder_cache_invalidation.md

```diff
@@ -0,0 +1,464 @@
+  redirect to the updated lease and that will also refresh the cache.
+
+**Pros:**
+
+- Addresses both [scenarios A and B](#scenarios).
+- By using gossip, the data will eventually reach every other
+  node even those that were temporarily unavailable.
+  (i.e. eliminates one cons of solution 1 above)
+- Does not add new data in gossip. Does not violate the [design
+  principles](#general-cockroachdb-design-principles).
+- No new data stored in the range cache to mark nodes
+  as poisoned.
+
+**Cons:**
+
+- Some complexity in cache lookups, to consult liveness
+  upon both cache hits and misses.
+- Added complexity during the lease discovery process.
+- Like above, there's a delay between the moment a node terminates
+  ([step 5 of drain](#draining)) and the moment other nodes learn of
+  the new lease records. During that time, requests can be mis-routed.
+
+  Like above, this last cons can be alleviated by extending step 5 of
+  the draining process by a "sleep period".
+
+
+### Multi-tenant variant
+
+- The KV servers receive the liveness updates over gossip from other
+  KV servers.
+- The KV servers *forward* the liveness updates to SQL tenants,
+  where the range caches are stored.
+
+## Summary of solutions and choice
+
+| Solution                                              | Addresses scenario A | Addresses scenario B | New data in gossip | New logic in cache         |
+|-------------------------------------------------------|----------------------|----------------------|--------------------|----------------------------|
+| [1](#1-cluster-wide-invalidation-via-RPC)             | yes                  | no                   | N/A                | no                         |
+| [2](#2-gossiping-lease-transfers)                     | yes                  | no                   | O(#leases)         | no                         |
+| [3](#3-gossiping-lease-acquisitions)                  | yes                  | yes                  | O(#ranges)         | no                         |
+| [4](#4-liveness-based-cache-invalidation-with-poison) | yes                  | yes                  | N/A                | Poisoned node/store IDs    |
+| [5](#4-liveness-based-cache-invalidation-with-bypass) | yes                  | yes                  | N/A                | Liveness bypass in lookups |
+
+Solutions 1 & 2 are undesirable becasue they don't address [scenario
+B](#scenarios).
+
+Solution 3 is undesirable because of the gossip storage requirement
+that goes against the [design
+principles](#general-cockroachdb-design-principles).
+
+Solutions 4 & 5 satisfy both scenarios without incurring extra gossip
+cost. However, they may require one additional cross-node hop for
+every range with outdated cache information, to discover the new
+lease. In contrast, solutions 2 & 3 pre-populate the caches with the
+location of the new lease.
+
+Solution 5 is simpler to implement than 4 but may incur lock
```

Comment:
I personally still stand in the Gossip camp :)
I think we'd benefit from caches being kept up to date under all conditions (well, maybe not under extreme conditions where there's too many updates to propagate), and I'm also interested in expanding a gossiping mechanism that I hope we introduce here to also deal with range splits/merges. Range descriptor changes are even more disruptive to traffic than lease changes, so I want to be proactive about them too.